contracts/contracts/strategies/NativeStaking/CompoundingValidatorManager.sol

```diff
@@ -0,0 +1,738 @@
+        IBeaconProofs(BEACON_PROOFS).verifyValidatorPubkey(
+            blockRoot,
+            consolidationLastPubKeyHash,
+            validatorPubKeyProof,
+            lastValidatorIndex
+        );
+
+        // Verify the balance of the last validator in the consolidation batch
+        // is zero. If its not then the consolidation has not been completed.
+        // This proof is to the beacon block root, not the balances container root.
+        uint256 validatorBalance = IBeaconProofs(BEACON_PROOFS)
+            .verifyValidatorBalance(
+                blockRoot,
+                balancesLeaf,
+                validatorBalanceProof,
+                lastValidatorIndex,
+                IBeaconProofs.BalanceProofLevel.BeaconBlock
+            );
+        require(validatorBalance == 0, "Last validator balance not zero");
+
+        // Reset the consolidation batch
+        consolidationLastPubKeyHash = bytes32(0);
+
+        // Take a snap of the balances so the new validator balances can be verified
+        _snapBalances();
+
+        // Do not unpause until after the new validator balances have been verified
+
+        // TODO emit an event
+    }
+
+    function snapBalances() public nonReentrant {
+        _snapBalances();
+    }
+
+    function _snapBalances() internal {
+        // Check no consolidation is waiting to be verified
+        require(
+            consolidationLastPubKeyHash == bytes32(0),
+            "Consolidation in progress"
+        );
+
+        bytes32 blockRoot = BeaconRoots.parentBlockRoot(
+            SafeCast.toUint64(block.timestamp)
+        );
+        // Get the current WETH balance
+        uint256 wethBalance = IWETH9(WETH).balanceOf(address(this));
+        // Get the current ETH balance
+        uint256 ethBalance = address(this).balance;
+
+        // Store the balances in the mapping
+        snappedBalances[blockRoot] = Balances({
+            timestamp: SafeCast.toUint64(block.timestamp),
+            wethBalance: SafeCast.toUint128(wethBalance),
+            ethBalance: SafeCast.toUint128(ethBalance)
+        });
+
+        // Store the snapped timestamp
+        lastSnapTimestamp = SafeCast.toUint64(block.timestamp);
+
+        emit SnappedBalances(
+            blockRoot,
+            block.number,
+            block.timestamp,
+            wethBalance,
+            ethBalance
+        );
+    }
+
+    struct VerifyBalancesParams {
+        bytes32 blockRoot;
+        uint64 firstPendingDepositSlot;
+        // BeaconBlock.BeaconBlockBody.deposits[0].slot
+        bytes firstPendingDepositSlotProof;
+        bytes32 balancesContainerRoot;
+        // BeaconBlock.state.validators
+        bytes validatorContainerProof;
+        bytes32[] validatorBalanceLeaves;
+        // BeaconBlock.state.validators[validatorIndex].balance
+        bytes[] validatorBalanceProofs;
+    }
+
+    function verifyBalances(VerifyBalancesParams calldata params)
```

Comment:
If I read this correctly it assumes we are always going to be able to verify the balances of all the validators the native staking contract holds?